x/storage/keeper/keeper_rate_limit.go

```diff
@@ -0,0 +1,232 @@
+	if found && currentRateLimit.FlowRateLimit.Equal(rateLimit) {
+		return nil
+	}
+
+	internalBucketInfo := k.MustGetInternalBucketInfo(ctx, bucketInfo.Id)
+
+	// if the rate limit is set to zero before, we should resume the payment of the bucket
+	if found && currentRateLimit.FlowRateLimit.IsZero() {
+		internalBucketInfo.PriceTime = ctx.BlockTime().Unix()
+
+		bill, err := k.GetBucketReadStoreBill(ctx, bucketInfo, internalBucketInfo)
+		if err != nil {
+			return fmt.Errorf("get bucket bill failed: %s %s", bucketInfo.BucketName, err.Error())
+		}
+		totalOutFlowRate := getTotalOutFlowRate(bill.Flows)
+		if totalOutFlowRate.GT(rateLimit) {
+			return fmt.Errorf("the total out flow rate(%s) of the bucket is greater than the new rate limit(%s)", totalOutFlowRate.String(), rateLimit.String())
+		}
+
+		return k.chargeBucketReadStoreFee(ctx, bucketInfo, internalBucketInfo)
+	}
+
+	// if the rate limit is not set t0 zero before, we should check the net flow rate of the bucket
+	bill, err := k.GetBucketReadStoreBill(ctx, bucketInfo, internalBucketInfo)
+	if err != nil {
+		return fmt.Errorf("get bucket bill failed: %s %s", bucketInfo.BucketName, err.Error())
+	}
+	totalOutFlowRate := getTotalOutFlowRate(bill.Flows)
+	if totalOutFlowRate.GT(rateLimit) {
+		return fmt.Errorf("the total out flow rate(%s) of the bucket is greater than the new rate limit(%s)", totalOutFlowRate.String(), rateLimit.String())
+	}
+
+	return nil
+}
+
+// getBucketFlowRateLimit returns the flow rate limit of the bucket from the store
+func (k Keeper) getBucketFlowRateLimit(ctx sdk.Context, paymentAccount sdk.AccAddress, bucketName string) (*types.BucketFlowRateLimit, bool) {
+	store := ctx.KVStore(k.storeKey)
+
+	bz := store.Get(types.GetBucketFlowRateLimitKey(paymentAccount, bucketName))
+	if bz == nil {
+		return nil, false
+	}
+
+	var rateLimit types.BucketFlowRateLimit
+	k.cdc.MustUnmarshal(bz, &rateLimit)
+	return &rateLimit, true
+}
+
+// setBucketFlowRateLimit sets the flow rate limit of the bucket to the store
+func (k Keeper) setBucketFlowRateLimit(ctx sdk.Context, paymentAccount sdk.AccAddress, bucketName string, rateLimit *types.BucketFlowRateLimit) {
+	store := ctx.KVStore(k.storeKey)
+
+	bz := k.cdc.MustMarshal(rateLimit)
+	store.Set(types.GetBucketFlowRateLimitKey(paymentAccount, bucketName), bz)
+}
+
+// isBucketFlowRateLimitSetToZero checks if the flow rate limit of the bucket is set to zero by the payment account owner
+func (k Keeper) isBucketFlowRateLimitSetToZero(ctx sdk.Context, paymentAccount sdk.AccAddress, bucketName string) bool {
+	rateLimit, found := k.getBucketFlowRateLimit(ctx, paymentAccount, bucketName)
+	if !found {
+		return false
+	}
+	return rateLimit.FlowRateLimit.IsZero()
+}
+
+// isBucketFlowRateUnderLimit checks if the flow rate of the bucket is under the flow rate limit
+func (k Keeper) isBucketFlowRateUnderLimit(ctx sdk.Context, paymentAccount, bucketOwner sdk.AccAddress, bucketName string, userFlows paymenttypes.UserFlows) error {
```

Comment:
This function can call `isBucketFlowRateUnderLimitWithRate` to reduce duplicated codes.